### PR TITLE
[NCCL] 2.19.4 build from source

### PR DESCRIPTION
**Pull request by simonbyrne**
It was suggested in https://github.com/JuliaPackaging/Yggdrasil/pull/7491#issuecomment-1753201161 to try building this from source, instead of repackaging the binaries.

### CODE REVIEW

**Comment by simonbyrne** [comment]
> catastrophic error: error while writing generated C file: No space left on device

Any ideas?

**Comment by giordano** [comment]
Print `df -hT` after the failure: https://github.com/JuliaPackaging/Yggdrasil/pull/7779#issuecomment-1846990383

**Comment by simonbyrne** [comment]
How do I run something after the failure?

**Comment by giordano** [comment]
```
command_that_fails || df -hT
```
like in https://github.com/JuliaPackaging/Yggdrasil/blob/4c71b6931aa043123b199dca953910b14fa41fa2/X/XLA/build_tarballs.jl#L27

**Comment by simonbyrne** [comment]
The non-x86_64 builds are failing because it is trying to run `fmt`: I added `coreutils` as a `BuildDependency`, but that installs the target arch version. Is there a way to get the x86_64 version?

**Comment by giordano** [comment]
HostBuildDependency: https://docs.binarybuilder.org/stable/build_tips/#Dependencies-for-the-target-system-vs-host-system

**Comment by simonbyrne** [comment]
https://buildkite.com/julialang/yggdrasil/builds/7094#018c5efb-c3a9-463c-9cc5-a1c47ab2f2e7/6-7609

**Comment by simonbyrne** [comment]
In the cross-compiled targets (aarch64 and powerpc64le) I get
```
[20:04:32] bash: line 1: /workspace/destdir/cuda/bin/nvcc: cannot execute binary file: Exec format error
```
Are we picking up the wrong `nvcc`?

**Comment by simonbyrne** [comment]
As for the space issue:
https://buildkite.com/julialang/yggdrasil/builds/7099#018c5f9f-b77a-41ed-8340-0095d6fbe0c7/6-2661
```
[20:06:02] ./prims_ll128.h(222): catastrophic error: error while writing generated C file: No space left on device
[20:06:02] 
[20:06:02] 1 catastrophic error detected in the compilation of "/workspace/destdir/obj/device/gensrc/reduce_scatter_sum_u8.cu".
[20:06:02] Compilation terminated.
```

and
https://buildkite.com/julialang/yggdrasil/builds/7099#018c5f9f-b77a-41ed-8340-0095d6fbe0c7/6-7520
```
[20:11:56]  ---> df -hT
[20:11:56] Filesystem           Type            Size      Used Available Use% Mounted on
[20:11:56] overlay              overlay         1.0G    184.0K   1023.8M   0% /
[20:11:56] df: /opt/x86_64-linux-musl/GCCBootstrap-4.8.5: No such file or directory
[20:11:56] df: /opt/x86_64-linux-musl/PlatformSupport-2023.6.10: No such file or directory
[20:11:56] df: /opt/x86_64-linux-musl/LLVMBootstrap-16.0.6: No such file or directory
[20:11:56] df: /opt/x86_64-linux-gnu/GCCBootstrap-4.8.5: No such file or directory
[20:11:56] df: /opt/x86_64-linux-gnu/PlatformSupport-2023.6.10: No such file or directory
[20:11:56] overlay              overlay         1.0G    184.0K   1023.8M   0% /opt/x86_64-linux-gnu
[20:11:56] overlay              overlay         1.0G    184.0K   1023.8M   0% /opt/x86_64-linux-musl
[20:11:56] udev                 devtmpfs      251.8G         0    251.8G   0% /dev/null
[20:11:56] udev                 devtmpfs      251.8G         0    251.8G   0% /dev/tty
[20:11:56] udev                 devtmpfs      251.8G         0    251.8G   0% /dev/urandom
[20:11:56] tank/root            zfs           228.6G     24.8G    203.8G  11% /etc/resolv.conf
[20:11:56] /dev/nvme0n1p1       ext4            3.4T      2.3T    938.6G  72% /root/.ccache
[20:11:56] tank/root            zfs           228.6G     24.8G    203.8G  11% /opt/toolchains
[20:11:56] tank/root            zfs           228.6G     24.8G    203.8G  11% /opt/bin
[20:11:56] /dev/nvme0n1p1       ext4            3.4T      2.3T    938.6G  72% /meta
[20:11:56] /dev/nvme0n1p1       ext4            3.4T      2.3T    938.6G  72% /workspace
```

so it doesn't look full?

**Comment by simonbyrne** [comment]
Okay, changing `TMPDIR` seemed to fix it.

Now I get https://buildkite.com/julialang/yggdrasil/builds/7100#018c6005-3d85-477f-9cbc-ca9413d67490/6-7523:
```
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/4.8.5/../../../../x86_64-linux-gnu/bin/ld: cannot find -lcudart_static
```

**Comment by simonbyrne** [comment]
Okay I think this is now working: @maleadt @vchuravy can you take a look?

Some questions:
- Can `nvcc` cross compile? I don't see any other examples where we do.
- Should I link against the `cudart_static` (the default) or `cudart` (what I've done above)?

**Comment by simonbyrne** [comment]
Also, should we delete the static library from the artifact? It's pretty big (200mb)

File: N/NCCL/build_tarballs.jl [hunk 1]
```diff
@@ -1,48 +1,58 @@
...
 
-cd ${WORKSPACE}/srcdir/nccl*
-find .
+export TMPDIR=${WORKSPACE}/tmpdir
+export CUDA_HOME=${WORKSPACE}/destdir/cuda
+export CUDA_LIB=${CUDA_HOME}/lib
+export CXXFLAGS='-D__STDC_FORMAT_MACROS'
+export CUDARTLIB=cudart
 
-install_license LICENSE.txt
```
Comment (imciner2) [inline N/NCCL/build_tarballs.jl:18]: We still need to keep the license in the build

**Review by imciner2 — CHANGES_REQUESTED** [review]

**Comment by maleadt** [comment]
> * Can `nvcc` cross compile? I don't see any other examples where we do.

No. You can try using clang for that, but I'm not sure it'll manage to compile all of NCCL's CUDA code.

> * Should I link against the `cudart_static` (the default) or `cudart` (what I've done above)?

The dynamic one is fine, and keeps the generated binaries smaller.

File: N/NCCL/build_tarballs.jl (hunk 1 above)
Comment (simonbyrne) [reply N/NCCL/build_tarballs.jl:18]: It's already in the tarball: is that sufficient?
```
├── LICENSE.txt
├── include
│   ├── nccl.h
│   └── nccl_net.h
├── lib
│   ├── libnccl.so -> libnccl.so.2
│   ├── libnccl.so.2 -> libnccl.so.2.19.4
│   ├── libnccl.so.2.19.4
│   ├── libnccl_static.a
│   └── pkgconfig
│       └── nccl.pc
└── share
    └── licenses
        └── NCCL
            └── LICENSE.txt
```

**Comment by simonbyrne** [comment]
Okay, I think this is ready. I'll leave the cross compilation part for now.

**Comment by simonbyrne** [comment]
Can we merge this?

**Comment by simonbyrne** [comment]
Hmm, it looks like it isn't loading correctly on Linux when no CUDA is installed:
```
(nccl) pkg> add NCCL_jll#main
...

julia> using NCCL_jll
ERROR: InitError: could not load library "/home/spjbyrne/.julia/artifacts/c99fd5bde3173b2e3f53a67a1383b5bb611c8267/lib/libnccl.so"
libcudart.so.12: cannot open shared object file: No such file or directory
Stacktrace:
  [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
    @ Base.Libc.Libdl ./libdl.jl:117
  [2] dlopen(s::String, flags::UInt32)
    @ Base.Libc.Libdl ./libdl.jl:116
  [3] macro expansion
    @ ~/.julia/packages/JLLWrappers/pG9bm/src/products/library_generators.jl:63 [inlined]
  [4] __init__()
    @ NCCL_jll ~/.julia/packages/NCCL_jll/qlpSf/src/wrappers/x86_64-linux-gnu-cuda+12.3.jl:10
  [5] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
    @ Base ./loading.jl:1115
  [6] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::String, depmods::Vector{Any})
    @ Base ./loading.jl:1061
  [7] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
    @ Base ./loading.jl:1506
  [8] _require(pkg::Base.PkgId, env::String)
    @ Base ./loading.jl:1783
  [9] _require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1660
 [10] macro expansion
    @ ./loading.jl:1648 [inlined]
 [11] macro expansion
    @ ./lock.jl:267 [inlined]
 [12] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1611
during initialization of module NCCL_jll
```

**Pull request merged**